In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# IMPORTANT: RUN THIS CELL IN ORDER TO IMPORT YOUR KAGGLE DATA SOURCES,
# THEN FEEL FREE TO DELETE THIS CELL.
# NOTE: THIS NOTEBOOK ENVIRONMENT DIFFERS FROM KAGGLE'S PYTHON
# ENVIRONMENT SO THERE MAY BE MISSING LIBRARIES USED BY YOUR
# NOTEBOOK.
import kagglehub
pypiahmad_realistic_action_recognition_ucf50_path = kagglehub.dataset_download('pypiahmad/realistic-action-recognition-ucf50')

print('Data source import complete.')

In [ ]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All"
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [ ]:
!pip install unrar

In [ ]:
!unrar x /content/drive/MyDrive/Advanced Computer Vision/Advanced_ComputerVision/Advanced/Lab - code/Lab10/UCF50.rar /content/


ERROR: Unknown option: 

In [ ]:
!mkdir -p "/content/UCF50_extracted"


In [ ]:
!unrar x "/content/drive/MyDrive/Advanced Computer Vision/Advanced_ComputerVision/Advanced/Lab - code/Lab10/UCF50.rar" "/content/UCF50_extracted/"


Kết quả truyền trực tuyến bị cắt bớt đến 5000 dòng cuối.
Extracting  /content/UCF50_extracted/UCF50/HorseRace/v_HorseRace_g16_c03.avi      23%  OK 
Extracting  /content/UCF50_extracted/UCF50/HorseRace/v_HorseRace_g16_c04.avi      23%  OK 
Extracting  /content/UCF50_extracted/UCF50/HorseRace/v_HorseRace_g16_c05.avi      23%  OK 
Extracting  /content/UCF50_extracted/UCF50/HorseRace/v_HorseRace_g17_c01.avi      23%  OK 
Extracting  /content/UCF50_extracted/UCF50/HorseRace/v_HorseRace_g17_c02.avi      23%  OK 
Extracting  /content/UCF50_extracted/UCF50/HorseRace/v_HorseRace_g17_c03.avi      23%  OK 
Extracting  /content/UCF50_extracted/UCF50/HorseRace/v_HorseRace_g17_c04.avi      23%  OK 
Extracting  /content/UCF50_extracted/UCF50/HorseRace/v_HorseRace_g17_c05.avi      23%  OK 
Extracting  /content/UCF50_extracted/UCF50/HorseRace/v_HorseRace_g18_c01.avi      23%  OK 
Extracting  /content/UCF50_extracted/UCF50/

###### Mỗi video sẽ được phân rã thành các frame, để trong cùng 1 thư mục có tên là phần tên video, các frame đánh chỉ số từ 00

In [ ]:
!mkdir -p "/content/UCF50_frames"

In [ ]:
import cv2
import os

# Đường dẫn tới thư mục UCF50
base_dir = '/content/UCF50_extracted/UCF50'

# Đường dẫn lưu khung hình
output_base_dir = '/content/UCF50_frames'

# Tạo thư mục lưu trữ nếu chưa tồn tại
if not os.path.exists(output_base_dir):
    os.makedirs(output_base_dir)

# Duyệt qua tất cả các thư mục trong UCF50
for category in os.listdir(base_dir):
    category_path = os.path.join(base_dir, category)

    # Kiểm tra xem có phải thư mục hay không
    if os.path.isdir(category_path):
        print(f"Đang xử lý thư mục: {category}")

        # Tạo thư mục cho từng category trong output
        category_output_dir = os.path.join(output_base_dir, category)
        if not os.path.exists(category_output_dir):
            os.makedirs(category_output_dir)

        # Duyệt qua các file video trong thư mục
        for video_file in os.listdir(category_path):
            if video_file.endswith('.avi'):  # Chỉ xử lý file .avi
                video_path = os.path.join(category_path, video_file)

                # Tạo thư mục cho từng video
                video_name = os.path.splitext(video_file)[0]
                video_output_dir = os.path.join(category_output_dir, video_name)
                if not os.path.exists(video_output_dir):
                    os.makedirs(video_output_dir)

                # Mở video
                vid = cv2.VideoCapture(video_path)
                frame_index = 0

                print(f"  Đang xử lý video: {video_file}")

                while True:
                    ret, frame = vid.read()
                    if not ret:
                        break  # Kết thúc video

                    # Lưu mỗi khung hình thứ 5
                    if frame_index % 5 == 0:
                        frame_name = os.path.join(video_output_dir, f'frame_{frame_index:04d}.jpg')
                        cv2.imwrite(frame_name, frame)

                    frame_index += 1

                vid.release()
                print(f"  Hoàn tất xử lý video: {video_file}")

print("Hoàn tất trích xuất tất cả các khung hình!")

Kết quả truyền trực tuyến bị cắt bớt đến 5000 dòng cuối.
  Đang xử lý video: v_Mixing_g25_c02.avi
  Hoàn tất xử lý video: v_Mixing_g25_c02.avi
  Đang xử lý video: v_Mixing_g14_c01.avi
  Hoàn tất xử lý video: v_Mixing_g14_c01.avi
  Đang xử lý video: v_Mixing_g14_c03.avi
  Hoàn tất xử lý video: v_Mixing_g14_c03.avi
  Đang xử lý video: v_Mixing_g13_c04.avi
  Hoàn tất xử lý video: v_Mixing_g13_c04.avi
  Đang xử lý video: v_Mixing_g11_c07.avi
  Hoàn tất xử lý video: v_Mixing_g11_c07.avi
  Đang xử lý video: v_Mixing_g01_c07.avi
  Hoàn tất xử lý video: v_Mixing_g01_c07.avi
  Đang xử lý video: v_Mixing_g08_c04.avi
  Hoàn tất xử lý video: v_Mixing_g08_c04.avi
  Đang xử lý video: v_Mixing_g07_c02.avi
  Hoàn tất xử lý video: v_Mixing_g07_c02.avi
  Đang xử lý video: v_Mixing_g11_c05.avi
  Hoàn tất xử lý video: v_Mixing_g11_c05.avi
  Đang xử lý video: v_Mixing_g08_c05.avi
  Hoàn tất xử lý video: v_Mixing_g08_c05.avi
  Đang xử lý video: v_Mixing_g10_c01.avi
  Hoàn tất xử lý video: v_Mixing_g10_c01.a

In [ ]:
import os
import shutil

# Đường dẫn đến thư mục UCF50_frames
ucf50_frames_path = '/content/UCF50_frames'

# Lặp qua tất cả các thư mục chính trong UCF50_frames
for main_folder in os.listdir(ucf50_frames_path):
    main_folder_path = os.path.join(ucf50_frames_path, main_folder)

    # Kiểm tra nếu đó là thư mục
    if os.path.isdir(main_folder_path):
        # Lấy danh sách các thư mục con
        subfolders = [f for f in os.listdir(main_folder_path) if os.path.isdir(os.path.join(main_folder_path, f))]

        # Tạo danh sách các thư mục cần giữ lại
        folders_to_keep = [
            folder for folder in subfolders
            if f"v_{main_folder}_g0" in folder and folder.split('_g0')[1][0].isdigit() and int(folder.split('_g0')[1][0]) <= 9
        ]

        # Xóa các thư mục không nằm trong danh sách cần giữ
        for folder in subfolders:
            if folder not in folders_to_keep:
                folder_path = os.path.join(main_folder_path, folder)
                shutil.rmtree(folder_path)
                print(f"Đã xóa: {folder_path}")

        print(f"Xử lý xong thư mục: {main_folder}")

print("Hoàn tất xử lý tất cả các thư mục.")

Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g19_c03
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g18_c01
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g22_c06
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g21_c04
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g24_c03
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g13_c05
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g17_c06
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g20_c01
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g15_c02
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g20_c06
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g20_c04
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g18_c04
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g15_c01
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g11_c04
Đã xóa: /content/UCF50_frames/PommelHorse/v_PommelHorse_g14_c04
Đã xóa: /content/UCF50_frames/PommelHors

In [ ]:
import tensorflow as tf
from tensorflow.keras import layers, Model, Input
import numpy as np
import os
import cv2
from sklearn.model_selection import train_test_split

def load_ucf50_data(base_dir='/content/UCF50_frames', num_frames=5, target_size=(150, 150)):
    """
    Load UCF50 frames from the specified directory
    Args:
        base_dir: Base directory containing UCF50 frames
        num_frames: Number of frames to load per video
        target_size: Size to resize frames to
    Returns:
        X: numpy array of frames
        y: numpy array of labels
    """
    X = []
    y = []
    label_map = {}

    # Get list of categories (classes)
    categories = sorted(os.listdir(base_dir))
    for i, category in enumerate(categories):
        label_map[category] = i
        category_path = os.path.join(base_dir, category)

        # Get list of video directories in this category
        video_dirs = os.listdir(category_path)

        for video_dir in video_dirs:
            video_path = os.path.join(category_path, video_dir)

            # Get list of frames in this video
            frames = sorted([f for f in os.listdir(video_path) if f.endswith('.jpg')])

            # Skip if not enough frames
            if len(frames) < num_frames:
                continue

            # Select evenly spaced frames
            selected_frames = np.linspace(0, len(frames)-1, num_frames, dtype=int)

            video_frames = []
            for frame_idx in selected_frames:
                frame_path = os.path.join(video_path, frames[frame_idx])

                # Read and preprocess frame
                frame = cv2.imread(frame_path)
                frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
                frame = cv2.resize(frame, target_size)
                frame = frame.astype('float32') / 255.0

                video_frames.append(frame)

            # Stack frames along channel dimension
            stacked_frames = np.concatenate(video_frames, axis=-1)

            X.append(stacked_frames)
            y.append(label_map[category])

    return np.array(X), np.array(y), label_map

def create_2dcnn_late_fusion(input_shape=(150, 150, 3), num_frames=5, num_classes=50):
    """
    Create 2D CNN model with Late Fusion
    Args:
        input_shape: Shape of each frame
        num_frames: Number of frames to process
        num_classes: Number of classes to predict
    """
    # Calculate input shape with all frames concatenated along channel dimension
    total_channels = input_shape[2] * num_frames
    fusion_input_shape = (input_shape[0], input_shape[1], total_channels)

    # Input layer
    inputs = Input(shape=fusion_input_shape)

    # Split input into separate frames
    frame_layers = []
    for i in range(num_frames):
        # Extract channels for current frame
        start_channel = i * 3
        frame = layers.Lambda(
            lambda x: x[:, :, :, start_channel:start_channel + 3]
        )(inputs)

        # First CNN block
        x = layers.Conv2D(16, (3, 3), activation='relu')(frame)
        x = layers.MaxPooling2D((2, 2))(x)
        x = layers.Dropout(0.3)(x)

        # Second CNN block
        x = layers.Conv2D(32, (3, 3), activation='relu')(x)
        x = layers.MaxPooling2D((2, 2))(x)
        x = layers.Dropout(0.2)(x)

        # Third CNN block
        x = layers.Conv2D(64, (3, 3), activation='relu')(x)
        x = layers.MaxPooling2D((2, 2))(x)
        x = layers.Dropout(0.1)(x)

        # Fourth CNN block
        x = layers.Conv2D(128, (3, 3), activation='relu')(x)
        x = layers.MaxPooling2D((2, 2))(x)

        frame_layers.append(x)

    # Late fusion using 1x1 convolution
    if len(frame_layers) > 1:
        concatenated = layers.Concatenate()(frame_layers)
        fused = layers.Conv2D(128, (1, 1), activation='relu')(concatenated)
    else:
        fused = frame_layers[0]

    # Flatten and MLP layers
    x = layers.Flatten()(fused)
    x = layers.Dense(256, activation='relu')(x)
    outputs = layers.Dense(num_classes, activation='softmax')(x)

    # Create model
    model = Model(inputs=inputs, outputs=outputs)

    # Compile model
    model.compile(
        optimizer='adam',
        loss='categorical_crossentropy',
        metrics=['accuracy']
    )

    return model

# Main training routine
if __name__ == "__main__":
    # Parameters
    NUM_FRAMES = 5
    TARGET_SIZE = (150, 150)
    NUM_CLASSES = 50
    BATCH_SIZE = 32
    EPOCHS = 50

    # Load data
    print("Loading data...")
    X, y, label_map = load_ucf50_data(
        base_dir='/content/UCF50_frames',
        num_frames=NUM_FRAMES,
        target_size=TARGET_SIZE
    )

    # Convert labels to one-hot encoding
    y = tf.keras.utils.to_categorical(y, NUM_CLASSES)

    # Split data
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42
    )

    # Create and compile model
    model = create_2dcnn_late_fusion(
        input_shape=TARGET_SIZE + (3,),
        num_frames=NUM_FRAMES,
        num_classes=NUM_CLASSES
    )

    # Print model summary
    model.summary()

    # Train model
    print("\nTraining model...")
    history = model.fit(
        X_train, y_train,
        batch_size=BATCH_SIZE,
        epochs=EPOCHS,
        validation_data=(X_test, y_test),
        callbacks=[
            tf.keras.callbacks.EarlyStopping(
                monitor='val_loss',
                patience=5,
                restore_best_weights=True
            )
        ]
    )

    # Evaluate model
    print("\nEvaluating model...")
    test_loss, test_acc = model.evaluate(X_test, y_test)
    print(f"Test accuracy: {test_acc:.4f}")

Loading data...


Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)  │ (None, 150, 150, 15)   │              0 │ -                      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ lambda (Lambda)           │ (None, 150, 150, 3)    │              0 │ input_layer[0][0]      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ lambda_1 (Lambda)         │ (None, 150, 150, 3)    │              0 │ input_layer[0][0]      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ lambda_2 (Lambda)         │ (None, 150, 150, 3)    │              0 │ input_layer[0][0]      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ lambda_3 (Lambda)         │ (None, 150, 150, 3)    │              0 │ input_layer[0][0]      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ lambda_4 (Lambda)         │ (None, 150, 150, 3)    │              0 │ input_layer[0][0]      │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv2d (Conv2D)           │ (None, 148, 148, 16)   │            448 │ lambda[0][0]           │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv2d_4 (Conv2D)         │ (None, 148, 148, 16)   │            448 │ lambda_1[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv2d_8 (Conv2D)         │ (None, 148, 148, 16)   │            448 │ lambda_2[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv2d_12 (Conv2D)        │ (None, 148, 148, 16)   │            448 │ lambda_3[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ conv2d_16 (Conv2D)        │ (None, 148, 148, 16)   │            448 │ lambda_4[0][0]         │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ max_pooling2d             │ (None, 74, 74, 16)     │              0 │ conv2d[0][0]           │
│ (MaxPooling2D)            │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ max_pooling2d_4           │ (None, 74, 74, 16)     │              0 │ conv2d_4[0][0]         │
│ (MaxPooling2D)            │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ max_pooling2d_8           │ (None, 74, 74, 16)     │              0 │ conv2d_8[0][0]         │
│ (MaxPooling2D)            │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ max_pooling2d_12          │ (None, 74, 74, 16)     │              0 │ conv2d_12[0][0]        │
│ (MaxPooling2D)            │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ max_pooling2d_16          │ (None, 74, 74, 16)     │              0 │ conv2d_16[0][0]        │
│ (MaxPooling2D)            │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dropout (Dropout)         │ (None, 74, 74, 16)     │              0 │ max_pooling2d[0][0]    │
├──────────────────────

 Total params: 2,187,986 (8.35 MB)

 Trainable params: 2,187,986 (8.35 MB)

 Non-trainable params: 0 (0.00 B)


Training model...
Epoch 1/50
58/58 ━━━━━━━━━━━━━━━━━━━━ 48s 425ms/step - accuracy: 0.0205 - loss: 3.9167 - val_accuracy: 0.0388 - val_loss: 3.8364
Epoch 2/50
58/58 ━━━━━━━━━━━━━━━━━━━━ 45s 79ms/step - accuracy: 0.0830 - loss: 3.6099 - val_accuracy: 0.1336 - val_loss: 3.4304
Epoch 3/50
58/58 ━━━━━━━━━━━━━━━━━━━━ 5s 88ms/step - accuracy: 0.2406 - loss: 2.8729 - val_accuracy: 0.2522 - val_loss: 2.9368
Epoch 4/50
58/58 ━━━━━━━━━━━━━━━━━━━━ 10s 79ms/step - accuracy: 0.4424 - loss: 2.0974 - val_accuracy: 0.3772 - val_loss: 2.3557
Epoch 5/50
58/58 ━━━━━━━━━━━━━━━━━━━━ 5s 90ms/step - accuracy: 0.6212 - loss: 1.4355 - val_accuracy: 0.4978 - val_loss: 1.9745
Epoch 6/50
58/58 ━━━━━━━━━━━━━━━━━━━━ 10s 79ms/step - accuracy: 0.7441 - loss: 0.9408 - val_accuracy: 0.5323 - val_loss: 1.9210
Epoch 7/50
58/58 ━━━━━━━━━━━━━━━━━━━━ 5s 86ms/step - accuracy: 0.7937 - loss: 0.7205 - val_accuracy: 0.5970 - val_loss: 1.6751
Epoch 8/50
58/58 ━━━━━━━━━━━━━━━━━━━━ 4s 75ms/step - accuracy: 0.8812 - loss: 0.4484 - 

In [ ]:
import tensorflow as tf
from tensorflow.keras import layers, Model, Input
import numpy as np
import os
import cv2
from sklearn.model_selection import train_test_split

def load_ucf50_data(base_dir='/content/UCF50_frames', num_frames=5, target_size=(150, 150)):
    """
    Load UCF50 frames from the specified directory with Early Fusion approach
    """
    X = []
    y = []
    label_map = {}

    # Get list of categories (classes)
    categories = sorted(os.listdir(base_dir))
    for i, category in enumerate(categories):
        label_map[category] = i
        category_path = os.path.join(base_dir, category)

        # Get list of video directories in this category
        video_dirs = os.listdir(category_path)

        for video_dir in video_dirs:
            video_path = os.path.join(category_path, video_dir)

            # Get list of frames in this video
            frames = sorted([f for f in os.listdir(video_path) if f.endswith('.jpg')])

            # Skip if not enough frames
            if len(frames) < num_frames:
                continue

            # Select evenly spaced frames
            selected_frames = np.linspace(0, len(frames)-1, num_frames, dtype=int)

            video_frames = []
            for frame_idx in selected_frames:
                frame_path = os.path.join(video_path, frames[frame_idx])

                # Read and preprocess frame
                frame = cv2.imread(frame_path)
                frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
                frame = cv2.resize(frame, target_size)
                frame = frame.astype('float32') / 255.0

                video_frames.append(frame)

            # Stack frames along channel dimension for Early Fusion
            stacked_frames = np.concatenate(video_frames, axis=-1)

            X.append(stacked_frames)
            y.append(label_map[category])

    return np.array(X), np.array(y), label_map

def create_2dcnn_early_fusion(input_shape=(150, 150, 15), num_classes=50):
    """
    Create 2D CNN model with Early Fusion
    Args:
        input_shape: Shape of input (height, width, 3*num_frames)
        num_classes: Number of classes to predict
    """
    # Input layer with 3*T channels
    inputs = Input(shape=input_shape)

    # First CNN block
    x = layers.Conv2D(16, (3, 3), padding='same', activation='relu')(inputs)
    x = layers.MaxPooling2D((2, 2))(x)
    x = layers.Dropout(0.3)(x)

    # Second CNN block
    x = layers.Conv2D(32, (3, 3), padding='same', activation='relu')(x)
    x = layers.MaxPooling2D((2, 2))(x)
    x = layers.Dropout(0.2)(x)

    # Third CNN block
    x = layers.Conv2D(64, (3, 3), padding='same', activation='relu')(x)
    x = layers.MaxPooling2D((2, 2))(x)
    x = layers.Dropout(0.1)(x)

    # Fourth CNN block
    x = layers.Conv2D(128, (3, 3), padding='same', activation='relu')(x)
    x = layers.MaxPooling2D((2, 2))(x)

    # Flatten and MLP layers
    x = layers.Flatten()(x)
    x = layers.Dense(256, activation='relu')(x)
    outputs = layers.Dense(num_classes, activation='softmax')(x)

    # Create model
    model = Model(inputs=inputs, outputs=outputs)

    # Compile model
    model.compile(
        optimizer='adam',
        loss='categorical_crossentropy',
        metrics=['accuracy']
    )

    return model

# Main training routine
if __name__ == "__main__":
    # Parameters
    NUM_FRAMES = 5
    TARGET_SIZE = (150, 150)
    NUM_CLASSES = 50
    BATCH_SIZE = 32
    EPOCHS = 100

    # Load data
    print("Loading data...")
    X, y, label_map = load_ucf50_data(
        base_dir='/content/UCF50_frames',
        num_frames=NUM_FRAMES,
        target_size=TARGET_SIZE
    )

    # Convert labels to one-hot encoding
    y = tf.keras.utils.to_categorical(y, NUM_CLASSES)

    # Split data
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=0.2, random_state=42
    )

    # Create and compile model
    model = create_2dcnn_early_fusion(
        input_shape=TARGET_SIZE + (3*NUM_FRAMES,),
        num_classes=NUM_CLASSES
    )

    # Print model summary
    model.summary()

    # Train model
    print("\nTraining model...")
    history = model.fit(
        X_train, y_train,
        batch_size=BATCH_SIZE,
        epochs=EPOCHS,
        validation_data=(X_test, y_test),
        callbacks=[
            tf.keras.callbacks.EarlyStopping(
                monitor='val_loss',
                patience=5,
                restore_best_weights=True
            )
        ]
    )

    # Evaluate model
    print("\nEvaluating model...")
    test_loss, test_acc = model.evaluate(X_test, y_test)
    print(f"Test accuracy: {test_acc:.4f}")

Loading data...


Model: "functional"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃ Layer (type)                         ┃ Output Shape                ┃         Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ input_layer (InputLayer)             │ (None, 150, 150, 15)        │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d (Conv2D)                      │ (None, 150, 150, 16)        │           2,176 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d (MaxPooling2D)         │ (None, 75, 75, 16)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout (Dropout)                    │ (None, 75, 75, 16)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_1 (Conv2D)                    │ (None, 75, 75, 32)          │           4,640 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_1 (MaxPooling2D)       │ (None, 37, 37, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout_1 (Dropout)                  │ (None, 37, 37, 32)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_2 (Conv2D)                    │ (None, 37, 37, 64)          │          18,496 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_2 (MaxPooling2D)       │ (None, 18, 18, 64)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dropout_2 (Dropout)                  │ (None, 18, 18, 64)          │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ conv2d_3 (Conv2D)                    │ (None, 18, 18, 128)         │          73,856 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ max_pooling2d_3 (MaxPooling2D)       │ (None, 9, 9, 128)           │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ flatten (Flatten)                    │ (None, 10368)               │               0 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense (Dense)                        │ (None, 256)                 │       2,654,464 │
├──────────────────────────────────────┼─────────────────────────────┼─────────────────┤
│ dense_1 (Dense)                      │ (None, 50)                  │          12,850 │
└──────────────────────────────────────┴─────────────────────────────┴─────────────────┘

 Total params: 2,766,482 (10.55 MB)

 Trainable params: 2,766,482 (10.55 MB)

 Non-trainable params: 0 (0.00 B)


Training model...
Epoch 1/100
58/58 ━━━━━━━━━━━━━━━━━━━━ 27s 281ms/step - accuracy: 0.0184 - loss: 4.0159 - val_accuracy: 0.0496 - val_loss: 3.8659
Epoch 2/100
58/58 ━━━━━━━━━━━━━━━━━━━━ 19s 49ms/step - accuracy: 0.0739 - loss: 3.6375 - val_accuracy: 0.1703 - val_loss: 3.3252
Epoch 3/100
58/58 ━━━━━━━━━━━━━━━━━━━━ 3s 45ms/step - accuracy: 0.2753 - loss: 2.7876 - val_accuracy: 0.3534 - val_loss: 2.4949
Epoch 4/100
58/58 ━━━━━━━━━━━━━━━━━━━━ 4s 71ms/step - accuracy: 0.5617 - loss: 1.7247 - val_accuracy: 0.5108 - val_loss: 2.0269
Epoch 5/100
58/58 ━━━━━━━━━━━━━━━━━━━━ 3s 49ms/step - accuracy: 0.7040 - loss: 1.1264 - val_accuracy: 0.5647 - val_loss: 1.7070
Epoch 6/100
58/58 ━━━━━━━━━━━━━━━━━━━━ 3s 49ms/step - accuracy: 0.7863 - loss: 0.7813 - val_accuracy: 0.6466 - val_loss: 1.4199
Epoch 7/100
58/58 ━━━━━━━━━━━━━━━━━━━━ 3s 49ms/step - accuracy: 0.8711 - loss: 0.4586 - val_accuracy: 0.6616 - val_loss: 1.3320
Epoch 8/100
58/58 ━━━━━━━━━━━━━━━━━━━━ 7s 78ms/step - accuracy: 0.8864 - loss: 0.3

## Đánh giá và nhận xét

Chú ý: Trước tiên thì do dữ liệu quá nhiều có thể dẫn đến việc training bị tràn RAM, lúc đó ta có thể phải cắt bớt dữ liệu đi. Ở mỗi thư mục, cụ thể trong thực nghiệm này, ta giữ lại các thư mục theo cú pháp: "v_tên thư mục_g0", tức là các thư mục từ g01 đến g09 sẽ được giữ lại, các thư mục còn lại thì sẽ bị xóa để thuận tiện cho việc training model.

**1. Kết quả Late Fusion:**

**Quá trình huấn luyện:**

Độ chính xác trên tập huấn luyện tăng dần và đạt mức rất cao (95.83% ở epoch 11 và 95.32% ở epoch 12), nhưng độ chính xác trên tập validation (val_accuracy) lại không đồng nhất và dao động thấp (chỉ đạt khoảng 60–63%).

Loss trên tập validation tăng liên tục qua các epoch (từ 1.8184 đến 2.2550).

**Đánh giá trên tập kiểm tra (test set):**

Độ chính xác trên tập kiểm tra (test accuracy) là 59.70%, thấp hơn nhiều so với huấn luyện.



**2. Kết quả Early Fusion:**
**Quá trình huấn luyện:**

Độ chính xác trên tập huấn luyện cũng rất cao, đạt 96.41% tại epoch 11.

Tuy nhiên, khác với Late Fusion, độ chính xác trên tập validation (val_accuracy) cải thiện tốt hơn, dao động từ 68.32% đến 73.28%, đồng thời val_loss cũng giảm dần (từ 1.3739 đến 1.3689 ở epoch 12).

**Đánh giá trên tập kiểm tra (test set):**

Độ chính xác trên tập kiểm tra là 66.16%, cao hơn so với Late Fusion.